app_test.go

```diff
@@ -860,45 +860,61 @@ func Test_App_ShutdownWithContext(t *testing.T) {
 	t.Parallel()
 
 	app := New()
+	shutdownHookCalled := false
+    app.Hooks().OnShutdown(func() error {
+        shutdownHookCalled = true
+        return nil
+    })
+
 	app.Get("/", func(ctx Ctx) error {
 		time.Sleep(5 * time.Second)
 		return ctx.SendString("body")
 	})
 
 	ln := fasthttputil.NewInmemoryListener()
 
-	go func() {
-		err := app.Listener(ln)
-		assert.NoError(t, err)
-	}()
-
-	time.Sleep(1 * time.Second)
-
-	go func() {
-		conn, err := ln.Dial()
-		assert.NoError(t, err)
-
-		_, err = conn.Write([]byte("GET / HTTP/1.1\r\nHost: google.com\r\n\r\n"))
-		assert.NoError(t, err)
-	}()
+    serverErr := make(chan error, 1)
+    go func() {
+        serverErr <- app.Listener(ln)
+    }()
+
+    time.Sleep(100 * time.Millisecond)
+
+    clientDone := make(chan struct{})
+    go func() {
+        conn, err := ln.Dial()
+        assert.NoError(t, err)
+        _, err = conn.Write([]byte("GET / HTTP/1.1\r\nHost: example.com\r\n\r\n"))
+        assert.NoError(t, err)
+        close(clientDone)
+    }()
+	
+	<-clientDone
+    time.Sleep(100 * time.Millisecond)
```

Comment:
why do we need a sleep here? i think it's arbitrary since we already wait for clientDone channel

Reply:
> why do we need a sleep here? i think it's arbitrary since we already wait for clientDone channel

This sleep is to ensure that our server has started processing the request. Although we have confirmed that the client has sent the request through <-clientDone, this means that the request has been sent to the server's listening port, and there is no guarantee that the server's processing coroutine has been sent. Start processing the request, so I think sleep 100 * Millisecond is required here as well.